openedx_tagging/core/tagging/api.py

```diff
@@ -98,5 +114,180 @@ def tag_object(
     Raised ValueError if the proposed tags are invalid for this taxonomy.
     Preserves existing (valid) tags, adds new (valid) tags, and removes omitted (or invalid) tags.
     """
-
     return taxonomy.tag_object(tags, object_id, object_type)
+
+
+def import_tags(taxonomy: Taxonomy, tags: BytesIO, format: TaxonomyDataFormat, replace=False):
+    """
+    Imports the hierarchical tags from the given blob into the Taxonomy.
+    The blob can be CSV or JSON format.
+
+    If replace, then removes any existing child Tags linked to this taxonomy before performing the import.
+    """
+
+    # Validations
+    if taxonomy.allow_free_text:
+        raise ValueError(
+            _(
+                f"Invalid taxonomy ({taxonomy.id}): You can't import free-from tags taxonomies"
+            )
+        )
+    if format not in TaxonomyDataFormat.__members__.values():
+        raise ValueError(
+            _(
+                f"Invalid format: {format}"
+            )
+        )
+
+    # Read file and build the tags data to be uploaded
+    try:
+        tags_data = {}
+        tags.seek(0)
+        if format == TaxonomyDataFormat.CSV:        
+            text_tags = TextIOWrapper(tags, encoding='utf-8')
+            csv_reader = csv.DictReader(text_tags)
+            header_fields = csv_reader.fieldnames
+            if csv_fields != header_fields:
+                raise ValueError(
+                    _(
+                        f"Invalid CSV header: {header_fields}. Must be: {csv_fields}."
+                    )
+                )
+            tags_data = list(csv_reader)
+        else:
+            # TaxonomyDataFormat.JSON
+            tags_data = json.load(tags)
+            if 'tags' not in tags_data:
+                raise ValueError(
+                    _(
+                        f"Invalid JSON format: Missing 'tags' list."
+                    )
+                )
+            tags_data = tags_data.get('tags')
+    except ValueError as e:
+        raise e
+    finally:
+        tags.close()
+
+
+    new_tags = []
+    updated_tags = []
+
+    def create_update_tag(tag):
+        """
+        Function to create a new Tag or update an existing one.
+
+        This function keeps a creation/update history with `new_tags` and `updated_tags`,
+        a same tag can't be created/updated in a same taxonomy import.
+        Also, recursively, creates the parents of the `tag`.
+
+        Returns the created/updated Tag.
+        Raise KeyError if 'id' or 'name' don't exist on `tag`
+        """
+
+        tag_id = tag['id']
+        tag_name = tag['name']
+        tag_parent_id = tag.get('parent_id')
+        tag_parent_name = tag.get('parent_name')
+
+        # Check if the tag has not already been created or updated
+        if tag_id not in new_tags and tag_id not in updated_tags:
+            try:
+                # Update tag
+                tag_instance = Tag.objects.get(external_id=tag_id)
```

Comment:
You also need to use `taxonomy=taxonomy` here, or you can have cross-taxonomy bugs.

Reply:
Oh wow.. I should have seen that!